feat: add new NotificationItem schema

mzikherman opened the pull request:
## What to do with this PR

This implements (more or less, some naming tweaks) the suggested schema in the tech plan - https://www.notion.so/artsy/Schema-for-Single-Notification-Page-9716fa45eee245d5b6def139d8c3d407.

There's no specs (yet), and we should discuss the naming and things like that, so please comment here!

## What is this

This introduces the `content: NotificationContent` schema under the `Notification` schema brought up in the [tech plan](https://www.notion.so/artsy/Schema-for-Single-Notification-Page-9716fa45eee245d5b6def139d8c3d407).

It allows a richer schema for each notification type to be specified (using a union type), so we can more easily build out a 'single-notification-landing-page' experience.

For instance, when someone lands/clicks/taps on an individual notification based on a viewing room being published -> you can specify `viewingRoom` (and whatever else we may add that is needed) in the fragment to render that page. When someone lands on a notification based on artworks matching an alert -> you can specify `artworksConnection` and `alert` (and whatever else we may add) in the fragment. And so forth.

I think this sets up development of richer landing pages for individual notifications / richer feeds, by allowing specific fragments for specific notification types to be specified w/ a sensible schema that makes sense for that notification type.

Additionally, it provides a pattern for adding new notification types in the future ('private offer received').

### Screenshots

**Artworks Published by Followed Artists** 
![Screen Shot 2023-12-08 at 3 38 04 PM](https://github.com/artsy/metaphysics/assets/1457859/41bc44ec-6477-40f3-80a3-c13b82749d8c)

**Artworks Matching Alert**
![Screen Shot 2023-12-08 at 3 37 21 PM](https://github.com/artsy/metaphysics/assets/1457859/ca100d7c-9e15-44b8-87c9-9addf052b8c8)

**Artist Featured in Article**
![Screen Shot 2023-12-08 at 3 50 19 PM](https://github.com/artsy/metaphysics/assets/1457859/b9f01f84-6058-4271-990a-b471fef3367b)

**Show Opened**
![Screen Shot 2023-12-08 at 3 55 01 PM](https://github.com/artsy/metaphysics/assets/1457859/04282d4f-4c2a-480e-83b8-ff4d79c71b56)

**Viewing Room Published**
![Screen Shot 2023-12-08 at 4 05 50 PM](https://github.com/artsy/metaphysics/assets/1457859/f94e4fdd-cea8-45df-903d-8c2542174255)

### Sample query

As mentioned in the tech plan, the client UI will likely include a fragment that looks like:

``` graphql
fragment NotificationLandingPage_content on NotificationContent {
  __typename
  ...ArtworkPublished_content
  ...CustomAlert_content
  ...ArticleFeaturedArtist_content
  # and so on for other supported types
}
```

**Here's how the schema in this PR satisfies that (long but this is what we need for all types to be supported!) The below query works for all existing notifications:**

``` graphql
{
  me {
    notification(id:"...") {
      internalID
      publishedAt(format: "RELATIVE")
      objectsCount

      content {
        __typename

        ... on ViewingRoomPublishedNotificationContent {
          viewingRoom {
            title
            partner {
              name
            }
          }
        }
        
        ... on ShowOpenedNotificationContent {
          show {
            name
            partner {
              ... on Partner {
                name
              }
            }
          }
        }
        
        ... on ArticleFeaturedArtistNotificationContent {
          article {
            title
          }
          
          artists {
            name
          }
        }
        
        ... on AlertNotificationContent {
          alert {
            additionalGeneNames
            attributionClass
            artists {
              name
            }
          }
          
          artworksConnection(first: 4) {
            edges {
              node {
                title
              }
            }
          }
        }
        
        ... on ArtworkPublishedNotificationContent {
          artists {
            name
          }

          artworksConnection(first: 4) {
            edges {
              node {
                title
              }
            }
          }
        }
      }
    }
  }
}
```

damassi commented on src/schema/v2/notifications/index.ts:
```diff
@@ -56,6 +57,10 @@ export const NotificationType = new GraphQLObjectType<any, ResolverContext>({
       deprecationReason: "Please use `publishedAt` instead",
     },
+    content: {
+      type: NotificationContentType,
```
notification types should live under the `/notifications/` directory

mzikherman replied:
True, the `content`, or whatever naming we come up with - can move to the `notifications/` directory.

damassi commented on src/schema/v2/notifications/index.ts:
```diff
@@ -56,6 +57,10 @@ export const NotificationType = new GraphQLObjectType<any, ResolverContext>({
       ...date(({ date }) => date),
       deprecationReason: "Please use `publishedAt` instead",
     },
+    content: {
```
```graphql
notification(id) {
  items {
    ... on ViewingRoomPublishedNotificationItem {
      ...
    }
  }
}
```

having the field be plural feels more flexible, since once can and will likely return multiple types. But at the same time it could also feel a bit confusing since one might assume a collection 🤷

mzikherman replied:
I'm not totally sold on this! It feels as though returning a single 'thing' that represents a single feed item (even though the feed item itself can be an aggregation of multiple 'activities' during a 24 hour period), is what we want.

damassi commented on src/schema/v2/NotificationContent/index.ts:
```diff
@@ -0,0 +1,178 @@
+  resolveType: ({ activity_type }) => {
+    switch (activity_type) {
+      case "SavedSearchHitActivity":
+        return AlertContentType
```
Why not match types 1:1 with their associated `activity_type`? That could solve the naming confusion. 

```
case "SavedSearchHitActivity": 
  return SavedSearchHitActivityType
```

mzikherman replied:
At least for `SavedSearch` naming, that's legacy naming and we prefer to always use `Alert` in the schema/front-end now, generally.

dblandin reviewed (commented):
This looks awesome! 💯

It's really nice to see these collaborating objects (artists, artworks, shows, viewing rooms) expressed more explicitly in the schema.

Here are some thoughts, which don't specifically refer to a particular line in the diff:

- Many of these notification types have the capability to bundle objects. For ViewingRoomPublishedActivity, it's possible to have multiple viewing rooms published by the same partner within the same time window and [expressed within the same aggregated feed event](https://github.com/artsy/gravity/blob/main/app/models/activity_feed/viewing_room_published_activity.rb#L23). Similar for shows published by the same partner [within the same time window](https://github.com/artsy/gravity/blob/main/app/models/activity_feed/partner_show_opened_activity.rb#L21). So whatever schema change ends up expressing viewing room / show objects, it would likely need to be of a plural type.
- For actors, most of the time, the events assume only one actor with the exception of ArtworkPublishedActivity which can express multiple artist actors associated with a bundled group of artworks published.

Given the above details of the current backend modeling, I wonder if it has been considered to recommit to the concept/abstraction of "actors" and "objects" in the schema, rather than model a new concept of "items" or "content" within the notification.

For example, you could have a request/response that looks something like that — introducing union types for both `actors` and `objects`, similar to the union type introduced within this PR for `content`:

**Example Request**

```graphql
{
  notification(id: "<ID>") {
    notificationType
    title
    message
    publishedAt

    actors(first: 5) { # `first` value is arbitrary for most events that assume only one actor but would be relevant for the ArtworkPublishedEvent currently.
      __typename
      ... on Alert {
        internalID
        name
      }
    }

    objects(first: 4) {
      __typename
      ... on Artwork {
        internalID
        title
      }
    }
  }
}
```

**Example Response**

```json
{
  "notification": { 
    "notificationType": "ARTWORK_ALERT",
    "title": "Works by Damien Hirst",
    "message": "1 work added",
    "publishedAt": "2023-12-08T20:04:39Z",

    "actors": [{
      "__typename": "Alert",
      "internalID": "356b3337-5b8e-48e7-b5bb-98f295c868c9",
      "displayName": "Damien Hirst — $0–$10,000, Painting or Print"
    }],

    "objects": [{
      "__typename": "Artwork",
      "internalID": "65748992b0faec000aeea49e",
      "title": "Norleucine",
    }]
  }
}
```

mzikherman commented:
Interesting @dblandin ! That's a good call out that these items can be aggregated, and so the singular versions within each type probably should be lists or connections.

I was only going by my feed (in production - staging essentially _has_ no feed), and my production feed did not have much (or any) aggregation happening. I wonder if that means we need a repeatable QA script that generates a robust set of aggregated feed items for a user, in order that we QA effectively. Otherwise, we'd be building a schema and working on the front-end w/o a full realistic set of data. This is a great example - I totally missed that events even _could_ be aggregated b/c that's not how my feed looks in production!

Re: `actors` and `objects` -> I'm mildly opinionated those are backend abstractions that shouldn't make it to the schema or clients. They are abstract, and it's not clear what they refer to (although GraphQL union types will help). Why call something `actor`, when it could be called `article`, or `artist` directly in the more specific notification type? I foresee people on the front-end flipping them around (or even here in the MP schema), and maybe even referring back to Gravity to double-check.

So I do think it's nice to forego that abstract naming, in favor of a friendlier and more convenient and explicit naming. Otherwise you'd need two union types for `actors` and `objects` , and....I don't know - I don't think we should have `actor` or `object` in our schema, really.

mzikherman commented on src/schema/v2/NotificationContent/index.ts:
```diff
@@ -0,0 +1,178 @@
+import {
```
Given the type naming for the parent `notification: Notification`, I would probably keep the `Notification`-named directory, and likely move `NotificationContent` to be sub-directories within there.

I'm 👍 on splitting up the individual types into their own file.

That leaves the naming of the sub-directory/type, trying to move away from `content` naming.

damassi replied:
Why don't we just go with `item` and keep it simple / idiomatic? 
- `ThingNotificationItem` 

```
item {
  ... on ThingNotificationItem {
    ... 
  }
}
```

mzikherman commented:
>Not sure if there's a way to make the Notification type itself a union type to surface these more explicit relationships without needing a nested content or body attribute.

@dblandin yup - the main cruftiness I still feel with this PR is where it lives in the schema. Working within a schema that _already_ has `Notification`, we're forced to come up with a totally new type `NotificationFeedItem`, or `NotificationContent`, and so forth, and expose that either as a new root field (or root field under `me`) alongside `notification`, _or_ as a field under the existing `Notification` (called `item` or `feedItem` or `content` or `detail`, etc.).

We should decide!

We have to avoid breaking changes to queries that target `Notification`, as that's being used in web + Eigen already, and I worry that updating it in-place to itself be a union type would be a breaking change.

Shall we update from `content: NotificationContent`, as a field under `Notification` -> to `feedItem: NotificationFeedItem`?

damassi commented:
`item` feels a little more general-purpose and less tied to the UI than `feedItem`, though if we renamed things up and down from `notifications` -> `feed` it might feel right.

mzikherman commented:
Thanks @dblandin and @damassi - updated with:

  - re-organization of files into an `Item/` sub-directory within `notifications/`
  - moved away from `content: NotificationContent` and to `item: NotificationItem`
  - added plural support via connections instead of a singular instance for certain types (h/t @dblandin)

damassi reviewed (approved):
Changes look good! 

One last one: It's a bit verbose, but for apps as big as MP verbosity tends to pay off. Can we name the file -> type 1:1, so 
```
- `AlertNotificationItem.ts`
- `ArticleFeaturedArtistNotificationItem.ts`
- etc
```
then cmd+p, search, etc, all yields precise entries, versus needing to derive information from the secondary context (parent folder name).

mzikherman commented:
Ah, interesting @damassi . Because of the verbosity and context provided by the parent directory I did intentionally keep that part off the filename. I think you're right though - the verbosity is actually _fine_ (don't need the parent dir. context), and sort of expected in a large/enterprise-y codebase.

anandaroop reviewed (commented):
Spike looks good to me! 

(Apart from Devon's caveats about `content.viewingRoom` etc needing to be plural).

In general I like having the more descriptive names vs "actors" and "objects", and assume those would be easier to work with on the front end.

Furthermore It's not clear to meet what the advantage would be of having actors and objects separated out in the response. I _think_ that while `UserAggregatedFeeds` _can_ bundle up multiple `Activities`, each instance will still only ever refer (via Activities) to a single "actor", due to [this](https://github.com/artsy/gravity/blob/5a08656cfe6ed3021d1fc3bb9ca232bf3a1d6dc4/app/models/activity_feed/aggregated_feed.rb#L49) reliance on group_key.

E.g. for `ArtworkPublishedActivity` feeds…

```ruby
artwork_published_feeds = UserAggregatedFeed.order(created_at: :desc).where(activity_type: 'ArtworkPublishedActivity').limit(10000).entries
#=> [ 10,000 recent "ArtworkPublished" feed entries ]

artwork_published_feeds.map{ |f| f.activities.map(&:actor_ids).uniq.length }.tally
#=> {1=>10000} # i.e. each of those 10,000 feeds refers to exactly one "actor"
```

So I guess this is a vote for a more UI-oriented `content` subfield, as opposed to carrying over some of these years-old abstractions from the backend which may not really line up with how we've ended up using them.

Let me know if I'm missed something about `feed <-> actor` cardinality though!

mzikherman commented on src/schema/v2/NotificationContent/index.ts:
```diff
@@ -0,0 +1,178 @@
+            totalCount,
+            pageCursors: createPageCursors({ page, size }, totalCount),
+            ...connectionFromArray(body, args),
+          }
```
You can likely use that helper here, but this is a bit of a special case where we have all the `object_ids` ahead of time, and so we always request the _entire_ connection, and then paginate as a convenience to the client.

Thus, since the entire list is already fetched, it feels more idiomatic to use `connectionFromArray` (which helps you slice the already-fetched array), vs `connectionFromArraySlice` - so I'm fine with this form, they'll both be equivalent.

mzikherman commented:
I've added specs here so I think it's mergeable! Will click the button so we can start trying to consume the schema on the FE and see how it feels.

cc @olerichter00 the schema changed a tiny bit!

joeyAghion commented:
The "actor" and "object" labeling wasn't so much a back-end abstraction as an attempt to generalize "activity" in a way that clients could easily render in a feed (a la Facebook), even if they didn't have custom logic for a particular type of activity. `<Someone>` is always doing something to `<some objects>`, or maybe multiple parties are doing the same thing to an object.

This alternative removes that generalization in favor of clients explicitly expecting specific activity types, which is more flexible and probably safer. (But I assume this results in errors if a client is unaware of a recently added activity type?)

MounirDhahri commented:
> But I assume this results in errors if a client is unaware of a recently added activity type?

Most likely yes, since it won't know how to deal with it. When querying for the `notificationsConnection` we will probably need to specify from the client side which types we would like to be returned to us.

mzikherman commented:
I think this more detailed schema is what would be used for a single-notification landing page, so a feed itself stays generic and probably would just work with a new type ('private offer received') added - although if there's an issue the new type could be filtered out by the API and require an opt-in query param to retrieve it.

For the single-notification landing page, if a user taps or clicks a notification type that isn't supported by the client, there'll be some fallback - today you usually get sent to the associated artwork page, or artist page - the generic schema has an `href` concept, so I imagine it should be pretty seamless for a client that _doesn't_ support a richer landing page for a notification type - just navigate the user to that `href` location.
